### PR TITLE
Added app

### PR DESCRIPTION
**Pull request by zralkasandra**

### CODE REVIEW

File: App.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        require('./assets/images/transportation.png'),
+        require('./assets/images/verify.png'),
+        require('./assets/images/chain.png'),
+      ]),
+      Font.loadAsync({
+        // This is the font that we are using for our tab bar
+        ...Icon.Ionicons.font,
+        // We include SpaceMono because we use it in HomeScreen.js. Feel free
+        // to remove this if you are not using it in your app
+        'space-mono': require('./assets/fonts/SpaceMono-Regular.ttf'),
```
Comment (tcichowicz) [inline App.js]: Seems like it's not used? Can we remove it?

File: App.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        'raleway': require('./assets/fonts/Raleway-SemiBold.ttf'),
+        'coiny': require('./assets/fonts/Coiny-Regular.ttf'),
+      }),
+    ]);
+  };
+
+  _handleLoadingError = error => {
+    // In this case, you might want to report the error to your error
+    // reporting service, for example Sentry
+    console.warn(error);
```
Comment (tcichowicz) [inline App.js:53]: In case of an error the app will display loading screen forever?

File: components/ScannedData.jsx [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    fontFamily: fonts.main,
+  },
+});
+
+const scannedData = (props) => {
+  const { data } = props;
+
+  const filteredData = Object.keys(data).map((item) => {
+    let row = null;
+    if (data[item] !== '') {
```
Comment (tcichowicz) [inline components/ScannedData.jsx]: Seems like a ternary operator (`condition ? val1 : val2`) would be a better fit here

File: package.json [hunk 4]
```diff
@@ -0,0 +1,39 @@
+{
+  "main": "node_modules/expo/AppEntry.js",
```
Comment (tcichowicz) [inline package.json:2]: Seems like we're missing a few key fields here such as name, version, description, author, license (MIT), repository, homepage, bugs - you can mostly copy those from the smart contracts repository.

File: screens/HomeScreen.jsx [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  static navigationOptions = {
+    header: null,
+  };
+
+  state = {
+    isVendor: '?',
+  }
+
+  componentDidMount() {
+    Web3Service.isVendor('0xeceffab2caf1ec535d407d366d747b575018d65e').then((res) => {
```
Comment (tcichowicz) [inline screens/HomeScreen.jsx]: Why do we have this here? Is this some kind of a visual debug?

File: screens/HomeScreen.jsx [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+              onPress={() => { this.onButtonPress('Verifier'); }}
+              text="Verify drug item"
+              img={verifyIcon}
+              buttonStyle="verify"
+              buttonTextStyle="verifyText"
+            />
+          </View>
+          <Text>
+            IsVendor:
+            {isVendor && isVendor.toString()}
```
Comment (tcichowicz) [inline screens/HomeScreen.jsx]: ditto

File: screens/QRCodeScreen.jsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+});
+
+export default class QRCodeScreen extends React.Component {
+  state = {
+    hasCameraPermission: null,
+  }
+
+  async componentDidMount() {
+    const { status } = await Permissions.askAsync(Permissions.CAMERA);
+    this.setState({ hasCameraPermission: status === 'granted' });
```
Comment (tcichowicz) [inline screens/QRCodeScreen.jsx]: Why do we have to repeat the same `hasCameraPermission` logic both here and in `QRCodeScanner.jsx`?

File: web3/service.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
+import Web3 from 'web3';
+import { abi } from './abi';
+
+const TEST_PROVIDER_URL = 'https://ropsten.infura.io/v3/da3111c149bb4134a9b3439392abe518';
```
Comment (tcichowicz) [inline web3/service.js]: This shouldn't really be committed, it's our private INFURA API key. Both this and `ADDRESS` could be in an `.env` file or parametrized in some other way good for React Native apps.

File: web3/service.js [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+  addAccount(account) {
+    this.web3.eth.accounts.wallet.add(account);
+  }
+
+  removeAccount(account) {
+    this.web3.eth.accounts.wallet.remove(account);
+  }
+
+  convertUtf8ToBytes(str) {
+    const shortVersion = str.replace(/-/g, '');
```
Comment (tcichowicz) [inline web3/service.js]: We should drop this replacing logic and check that the string has 32 characters.

File: web3/service.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+  getGasPrice() {
+    return this.web3.eth.getGasPrice();
+  }
+
+  getBlock(blockNo) {
+    return this.web3.eth.getBlock(blockNo);
+  }
+
+  getTransactionCount() {
```
Comment (tcichowicz) [inline web3/service.js]: Some of these functions are unused, can we remove them?

File: web3/service.js [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+  async verifyItem(drugItemId) {
+    const drugItemBytes = this.convertUtf8ToBytes(drugItemId);
+    let result = '';
+    try {
+      const codes = await this.contract.methods
+        .isPurchasable(drugItemBytes)
+        .call();
+      const uniqueCodes = [...new Set(codes)];
+
+      uniqueCodes.forEach((code) => {
```
Comment (tcichowicz) [inline web3/service.js]: Map or reduce would be a much better fit here :)

File: App.js (hunk 1 above)
Comment (zralkasandra) [reply App.js]: Yes

File: components/ScannedData.jsx (hunk 3 above)
Comment (zralkasandra) [reply components/ScannedData.jsx]: ok

File: package.json (hunk 4 above)
Comment (zralkasandra) [reply package.json:2]: ok

File: screens/HomeScreen.jsx (hunk 5 above)
Comment (zralkasandra) [reply screens/HomeScreen.jsx]: I used that to see if I everything with web3js is fine. I can remove it now.

File: screens/QRCodeScreen.jsx (hunk 7 above)
Comment (zralkasandra) [reply screens/QRCodeScreen.jsx]: This file should be removed

File: web3/service.js (hunk 9 above)
Comment (zralkasandra) [reply web3/service.js]: ok

File: web3/service.js (hunk 10 above)
Comment (zralkasandra) [reply web3/service.js]: ok

File: web3/service.js (hunk 11 above)
Comment (zralkasandra) [reply web3/service.js]: ok

File: web3/service.js (hunk 8 above)
Comment (zralkasandra) [reply web3/service.js]: ok

File: App.js [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        require('./assets/images/robot-prod.png'),
+        require('./assets/images/register.png'),
+        require('./assets/images/transportation.png'),
+        require('./assets/images/verify.png'),
+        require('./assets/images/chain.png'),
+      ]),
+      Font.loadAsync({
+        // This is the font that we are using for our tab bar
+        ...Icon.Ionicons.font,
+        // We include SpaceMono because we use it in HomeScreen.js. Feel free
```
Comment (tcichowicz) [inline App.js]: Obsolete comment and the font file is still there as well.

File: components/ScannedData.jsx [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+          <Text style={styles.label}>
+            {`${item}:`}
+          </Text>
+          <Text style={styles.value}>
+            {data[item]}
+          </Text>
+        </View>
+      )
+      : null;
+    return row;
```
Comment (tcichowicz) [inline components/ScannedData.jsx:71]: You can use a short lambda without curly braces if you get rid of the `row` variable.

File: web3/service.js [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+  async verifyItem(drugItemId) {
+    const drugItemBytes = this.convertUtf8ToBytes(drugItemId);
+    let result = '';
+    try {
+      const codes = await this.contract.methods
+        .isPurchasable(drugItemBytes)
+        .call();
+      const uniqueCodes = [...new Set(codes)];
+
+      result = uniqueCodes.map((code) => {
```
Comment (tcichowicz) [inline web3/service.js]: Wouldn't just `uniqueCodes.map(this.drugStates.get).join('.')` work? BTW why join with a dot?

**Review by tcichowicz — APPROVED** [review]

**Pull request merged**